### ISSUE TITLE
[RFC] Replace Void with Nil

### ISSUE THREAD

**Issue by asterite** [crystal-lang/crystal#2673]
Right now the language has a `Void` type. It came to existence when we added C bindings to the language: some functions don't return anything, `void` in C, so we added a similar type to Crystal.

The problem is, `Void` doesn't fit nicely into the language. Crystal is a language where every expression has a value, and `Void` means "has no value". For example, you can't print a `Void` value, and you can't add methods to it, because it has no methods at all.

We already have a type that means "absence of value": `Nil`

So, we could replace `Void` with `Nil`. A lib function that returns `Void` would now return `Nil`:

```crystal
lib LibFoo
  fun foo
end

puts LibFoo.foo # => "nil"
```

We could still have an `alias Void = Nil`, mostly because when writing some code that interacts with C one would like to write `Void*`, or simply say `fun foo : Void`. That's fine, but from Crystal's side it would still mean `nil`.

The codegen would make methods and functions returning `nil` as returning `void`.

We could even add the rule that if you mark a method as returning `Nil`, the return value doesn't matter, so this compiles:

```crystal
# Means: I don't care about the return value
def foo : Nil
  do_something # maybe this returns an Int32, but this will still compile
end

foo # => nil
```

This last part is somewhat related to #1393, where `nil` was also suggested as a more intuitive type than `Void`, which is a special case.

**Comment by lbguilherme** [comment]
I agree with this change, but: Would `GC.malloc` now return a `Nil*`? That seems a bit odd since while `Void` is meant to be the lack of value, a pointer to void is viewed as a pointer to "anything". I don't think `Nil*` translates well here.

**Comment by kirbyfan64** [comment]
Why not just have `malloc` return a pointer to bytes? I mean, really that's what it is...

**Comment by asterite** [comment]
Good points. I think instead of `Void*` we should use `UInt8*`, basically a pointer to bytes, as @kirbyfan64 says.

**Comment by bcardiff** [comment]
Nil is not a 100% replacement of Void. In union types they do have different semantic.
  * `Nil | T = T?`
  * `Void | T = Void`

So, `typeof( cond ? something_that_returns_void : expr ) # => Void` or any other conditional should hide the other branch value in order to make a sound result.

If we use `: Nil` return type in methods to swallow intermediate results i feel like we are giving more roles to `Nil` than what it should.

But I understand is something pragmatic :-) so 👍

**Comment by asterite** [comment]
@bcardiff Almost, we are still missing that rule in the language (`Void | T` = `Void`). I think that can also apply for procs with the same argument types but one returning Void and other not (#2672):

```
Proc(A, B, Void) | Proc(A, B, T) = Proc(A, B, Void)
```

I'll first implement these rules, which make sense, and then come back to this issue. I'm still not 100% sure about removing `Void`.

**Comment by bcardiff** [comment]
`Proc(A, B, S) | Proc(A, B, T) = Proc(A, B, S | T)`
from the user perspective. That is the rule you want from procs. Not one with Void. Together with `Void | T = Void`, you get the equivalence you mention.

**Comment by asterite** [comment]
@bcardiff The problem is that if we have `Proc(A, B, S | T)` but the real proc behind that value is `Proc(A, B, S)`, then when we invoke it we have no idea that we would need to convert that `S` to an `S | T`. Basically, it can be done because codegen/binary problems. On the other hand, `Proc(A, B, S | T)` and `Proc(A, B, S) | Proc(A, B, T)` are indistinguishable, and they both accept the same arguments and return the same type, so it's OK to keep the union.

**Comment by asterite** [comment]
Fixed by #2701
